### DISCUSSION TITLE
/react-vscode-webview-hot-module-replacement/

### DISCUSSION THREAD

**Discussion by comment-migrator** [estruyf/eliostruyf-comments#588] [Comments]
Use React in your VSCode WebView with hot module replacement
        
[https://www.eliostruyf.com/react-vscode-webview-hot-module-replacement/](https://www.eliostruyf.com/react-vscode-webview-hot-module-replacement/)

**Comment by comment-migrator** [comment]
```header
Originally posted by Paweł Syska on 2022-02-03 21:08
```

Hello, I managed to display the content of the localhost React page, but I'm getting WebSocket connection failed error instantly (error in the Developer Tools console). I've basically copy-pasted the webpack configuration from your Frontmatter extension but it didn't change anything. Do you have any idea why this could happen? ![https://uploads.disquscdn.c...](https://uploads.disquscdn.com/images/65296775bf859e583ec63edf83c6921e1055715291c988e9a251fb274fa467b8.png)

**Reply by comment-migrator** [reply]
```header
Originally posted on 2022-02-04 08:37
```

@pawesyska:disqus you need to specify the CSP for development, that will probably the reason why it gets blocked.

Here is how it's done in Front Matter: [https://github.com/estruyf/...](https://github.com/estruyf/vscode-front-matter/blob/main/src/commands/Dashboard.ts#L203 "https://github.com/estruyf/vscode-front-matter/blob/main/src/commands/Dashboard.ts#L203")

Above, there was no CSP yet defined. That is why no issues occured.

**Comment by phil294** [comment]
Thanks for writing this up, this helped me to enable HMR for my Vue.js 3 based webview extension. Hoping it may be useful to someone else in the future, I want to leave a link showing all my changes that were necessary: https://github.com/phil294/git-log--graph/commit/c050f714876c7032d8c66157160edaf3a06f2127

Edit: This overwrites the `publicPath` and if you need to access `process.env.BASE_URL`, you're out of luck in production, as the path is non-stable. The way I solved this is by looking up the path by looking at the `<script src="...">` of your webview markup if necessary for dynamic assets loading: https://github.com/phil294/git-log--graph/commit/d72cf69b38e816127b55363cf4f25b59c5c247ab

**Reply by estruyf** [reply]
Thanks @phil294 for sharing! 🙏

**Comment by Nash171** [comment]
Hi, I learned a lot from your articles and codes, Thanks for these great stuff :)

This question not directly related to this article, but I couldn't find better article to ask this. 
I wrote a Custom Editor WebView extension using React and its working well.
Now I'm trying to integrate MSW's browser integration in my React app, but no success yet.

Having issues with loading MSW's service worker from webview, and I'm still trying to figure it out.

Do you think its possible to use MSW in WebView React app?

**Reply by estruyf** [reply]
Hi @Nash171, thanks for the feedback.

I've never used MSW, but I quickly looked into it, and I see that it gets stuck when loading the service worker because it runs on a different origin.

<img width="925" alt="image" src="https://github.com/estruyf/eliostruyf-comments/assets/2900833/8c389f41-5378-48a1-8c09-a9ce2dd4fe4b">

Not sure if there is a way to manually register the service worker and allow MSW to use it.

**Reply by Nash171** [reply]
Thanks for the response. Yes its runs on a different origin, I couldn't resolve it yet :(

There is a way to pass service worker url when starting worker, but I still think it doesn't solve the origin issue.
https://mswjs.io/docs/api/setup-worker/start#url

Thank you for giving it a try.

**Reply by estruyf** [reply]
Nope, I had also tried that; the problem is that we cannot link any files from the web view URL, and the MSW library expects this.
